### ISSUE TITLE
Some uses of #[derive_where(Clone, Copy)] trigger clippy lints

### ISSUE THREAD

**Issue by eric-seppanen** [ModProg/derive-where#91]
Example:
```rust
#[derive_where(Clone, Copy; F:Copy)]
pub struct Thing<T, F> {
    index: F,
    _marker: PhantomData<fn() -> T>,
}
```
This triggers the [non_canonical_clone_impl](https://rust-lang.github.io/rust-clippy/rust-1.74.0/index.html#/non_canonical_clone_impl) clippy lint:

```txt
warning: non-canonical implementation of `clone` on a `Copy` type
 --> src/lib.rs:7:1
  |
7 | #[derive_where(Clone, Copy; F:Copy)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
```

**Comment by eric-seppanen** [comment]
The expanded code looks like this:
```rust
impl<T, F> ::core::clone::Clone for Thing<T, F>
where
    F: Copy,
{
    #[inline]
    fn clone(&self) -> Self {
        match self {
            Thing { index: ref __field_index, _marker: ref __field__marker } => {
                Thing {
                    index: ::core::clone::Clone::clone(__field_index),
                    _marker: ::core::clone::Clone::clone(__field__marker),
                }
            }
        }
    }
}
impl<T, F> ::core::marker::Copy for Thing<T, F>
where
    F: Copy,

```

It appears (from viewing the expanded code) that if I do `derive_where(Clone, Copy)` I get the `Clone` impl that clippy expects; but if I add the trait bound (`derive_where(Clone, Copy; F:Copy)`) I get the explicit impl that makes clippy sad.

**Comment by daxpedda** [comment]
I see.
For hygiene reasons we probably shouldn't try to detect `F: Copy`, because we can't make sure that `Copy` is `core::marker::Copy`.

In any case, this is a bug in Clippy, as this lint should probably not be triggered in proc-macros. I don't know how we stand on adding a `#[allow(clippy::non_canonical_clone_impl)]` in the meantime.

Cc @ModProg.

**Comment by daxpedda** [comment]
Well, interesting: https://github.com/rust-lang/rust-clippy/issues/11080.

Maybe we should apply `#[automatically_derived]` in `derive-where`?

**Comment by eric-seppanen** [comment]
> For hygiene reasons we probably shouldn't try to detect F: Copy, because we can't make sure that Copy is core::marker::Copy.

I don't understand why the trait bounds matter.

The macro itself is generating `impl Copy for Thing`, so I think that the `impl Clone for thing` that defers to the `Copy` should always be possible.

Even if there are other ways to suppress the lint, it seems like the right thing to do as it may generate more efficient code (i.e. memcpy rather than a field-by-field clone.)

**Comment by daxpedda** [comment]
Yeah, indeed these two problems are unrelated.

> (as long as the trait bounds are identical for both)

Seems correct to me.

**Comment by daxpedda** [comment]
Thanks @eric-seppanen, I addressed this in #95.

**Comment by eric-seppanen** [comment]
I have also been confused by the documentation for `#[automatically_derived]`. I opened https://github.com/rust-lang/reference/issues/1439 to see if the rust reference is the right place to cover this.
